ppafm/ml/Generator.py

```diff
@@ -389,29 +395,33 @@ def __init__(
         self.df_steps = self.afmulator.df_steps
         self.z_size = self.scan_dim[2] - self.df_steps + 1
 
-    def _prepareBuffers(self, rhos=None, rho_deltas=None):
+    def _prepare_tip_buffers(self, rhos=None, rho_deltas=None, density_cutoff=None):
         if rhos is None:
-            self.rhos = self.ffts = [(None, None)] * len(self.iZPPs)
+            self.rhos_tip = self.rhos_tip_delta = self.ffts_tip = self.ffts_tip_delta = [None] * len(self.iZPPs)
             return
         elif len(rhos) != len(self.iZPPs):
             raise ValueError(f"The length of rhos ({len(rhos)}) does not match the length of iZPPs ({len(self.iZPPs)})")
         if rho_deltas is not None and (len(rhos) != len(rho_deltas)):
             raise ValueError(f"The length of rhos ({len(rhos)}) does not match the length of rho_deltas ({len(rho_deltas)})")
-        self.rhos = []
-        self.ffts = []
-        for i in range(len(rhos)):
-            self.afmulator.setRho(rhos[i], sigma=self.afmulator.sigma, B_pauli=self.afmulator.B_pauli)
-            rhos_ = [self.afmulator.forcefield.rho]
-            ffts_ = [self.afmulator.forcefield.fft_corr]
-            if rho_deltas is not None:
-                self.afmulator.setRhoDelta(rho_deltas[i])
-                rhos_.append(self.afmulator.forcefield.rho_delta)
-                ffts_.append(self.afmulator.forcefield.fft_corr_delta)
-            else:
-                rhos_.append(None)
-                ffts_.append(None)
-            self.rhos.append(rhos_)
-            self.ffts.append(ffts_)
+        self.rhos_tip = []
+        self.ffts_tip = []
+        self._rhos_original = []
```

Comment:
Why `_rhos_original` is necessary?

Reply:
It is used when changing the FDBM parameters, specifically the exponent `B_pauli`. When the exponent changes, the array needs to be prepared again starting from the original array.